### PR TITLE
[WEEX-233][iOS] More enhanced about <web> component

### PR DESCRIPTION
**Pull request by tw93**
Hi, I'm Tw93(侑夕) from [Fliggy](https://www.fliggy.com/), I'd like to discuss the enhancements about [<Web>](https://weex.apache.org/references/components/web.html) component, let it take over where we can't implement through weex.

## Target

`<web>` is used to display web content that specified by src attribute in weex page,We also can use webview module to control WebView behavior such as goBack, goForward and reload.

But it's not enough in most businesses，I think a more enhanced web component should have the following:

* Support to send messages from Weex to a html in <web> component.
* Support to send messages from a html in <web> component to Weex.
* Support to render html source.

Previously, <web> component was an island that only rendered a remote url, but enabled it to communicate through the context native to the webView. Then I'm going to describe the solution that I came up with.

### Code
* Weex iOS:
  * [WXWebComponent.m](https://github.com/tw93/incubator-weex/blob/ios-feature-enhanced-web/ios/sdk/WeexSDK/Sources/Component/WXWebComponent.m)
  * [WXWebComponent.h](https://github.com/tw93/incubator-weex/blob/ios-feature-enhanced-web/ios/sdk/WeexSDK/Sources/Component/WXWebComponent.h)
  * [WXWebViewModule.m](https://github.com/tw93/incubator-weex/blob/ios-feature-enhanced-web/ios/sdk/WeexSDK/Sources/Module/WXWebViewModule.m)
* Weex Vue Demo:
  * [http://dotwe.org/vue/3576ed477d21dc020e63bf57dc414fff](http://dotwe.org/vue/3576ed477d21dc020e63bf57dc414fff)
  * [Demo Bundle JS](http://dotwe.org/raw/dist/3576ed477d21dc020e63bf57dc414fff.bundle.wx)
  * [Web End](http://h5.m.taobao.com/trip/wx-detection-demo/web.html)

### [Demo](https://gw.alipayobjects.com/zos/rmsportal/sBfTGamezxaBCGPshyXT.gif) && [Theory](https://img.alicdn.com/tfs/TB1hEhXbxGYBuNjy0FnXXX5lpXa-1872-1208.png)

<table><tr><td></td><td></td></tr></table>

#### Support to send messages from Weex to a html in <web> component.

* Native：(W3c [MessageEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent))
 
  ```
   // Weex postMessage to web
  - (void)postMessage:(NSDictionary *)data {
      WXSDKInstance *instance = [WXSDKEngine topInstance];
  
      NSMutableString *bundleUrlOrigin = @"";
  
      if (instance.pageName) {
          NSString *bundleUrl = [instance.scriptURL absoluteString];
          NSURL *url = [NSURL URLWithString:bundleUrl];
          bundleUrlOrigin = [NSString stringWithFormat:@"%@://%@%@", url.scheme, url.host, url.port ? [NSString stringWithFormat:@":%@", url.port] : @""];
      }
  
      NSDictionary *initDic = @{
          @"type" : @"message",
          @"data" : data,
          @"origin" : bundleUrlOrigin
      };
  
      NSString *json = [WXUtility JSONString:initDic];
  
      NSString *code = [NSString stringWithFormat:@"(function (){window.dispatchEvent(new MessageEvent('message', %@));}())", json];
      [_jsContext evaluateScript:code];
  }
  ```
  
* How to use：
 
  ```
    // Weex
    const webview = weex.requireModule('webview');
    // recommend
    this.$refs['wxc-web'].postMessage({ detail: "a message" });
    //not recommend
    //webview.postMessage(this.$refs.webview, {detail:"a message"});
  
    // Web 
    window.addEventListener('message',function(e){
    console.log(e.data,e.type,e.origin); // { detail: "a message" },"message","*"
  },false)
  ```

#### Support to send messages from a html in <web> component to Weex.

- Native：(W3c [MessageEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent))

  ```
  //Weex catch postMessage event from web
      _jsContext[@"postMessage"] = ^() {
  
          NSArray *args = [JSContext currentArguments];
  
          if (args && args.count < 2) {
              return;
          }
  
          NSDictionary *data = [args[0] toDictionary];
          NSString *origin = [args[1] toString];
  
          if (data == nil) {
              return;
          }
  
          NSDictionary *initDic = @{ @"type" : @"message",
                                     @"data" : data,
                                     @"origin" : origin
          };
  
          [weakSelf fireEvent:@"message" params:initDic];
      };
  ```

- How to use：

  ```
    // Weex
    <web @message="onMessage"></web>
    
     onMessage (e) {
            console.log(e.data,e.type,e.origin); // { detail: "a message" },"message","*"
      },
    
    // Web iframe
      window.parent.postMessage({ detail: "a message" }, '*');
  ``` 

#### Support to render html source.

* Native:
 
  ```
  [_webview loadHTMLString:_source baseURL:nil];
  ```
* How to use:
 
  ```html
  <web source='<p style="text-align: center;"><img align="absmiddle" src="http://img03.taobaocdn.com/imgextra/i3/1124701655/TB2zmmIcpXXXXbIXpXXXXXXXXXX_!!1124701655.jpg"/></p>'></web>
  ```

Welcome to put forward any suggestion about the solution or other requirements for the <web> component, Thanks！

### CODE REVIEW

File: ios/sdk/WeexSDK/Sources/Component/WXWebComponent.m [hunk 1]
```diff
@@ -69,6 +71,11 @@ - (instancetype)initWithRef:(NSString *)ref type:(NSString *)type styles:(NSDict
 {
     if (self = [super initWithRef:ref type:type styles:styles attributes:attributes events:events weexInstance:weexInstance]) {
         self.url = attributes[@"src"];
+        
+        if(attributes[@"source"]){
+            self.source = attributes[@"source"];
```
Comment (acton393) [inline ios/sdk/WeexSDK/Sources/Component/WXWebComponent.m]: the init method is executed in background thread named component method and  `self.source` will call [self setSource:""], 
the method for `setSource` called view operation which is only for main thread.

A better way is that define another member for source to save the variable during  this initialization,  and call  self.source in component hook lifeCycle like  `viewDidLoad`.

File: ios/sdk/WeexSDK/Sources/Module/WXWebViewModule.m [hunk 2]
```diff
@@ -27,6 +27,7 @@ @implementation WXWebViewModule
 @synthesize weexInstance;
```
[the remark this answers is not in the document]
Comment (tw93) [reply ios/sdk/WeexSDK/Sources/Module/WXWebViewModule.m]: Thinks @acton393 , I have modified it

File: ios/sdk/WeexSDK/Sources/Component/WXWebComponent.m [hunk 3]
```diff
@@ -104,6 +142,10 @@ - (void)updateAttributes:(NSDictionary *)attributes
     if (attributes[@"src"]) {
         self.url = attributes[@"src"];
     }
+
+    if (attributes[@"source"]) {
```
Comment (acton393) [inline ios/sdk/WeexSDK/Sources/Component/WXWebComponent.m]: maybe we should update the initSource when update attributes,
as the view can be unloaded during the screen scroll, so the viewDidLoad method can called again with initSource

Comment (tw93) [reply ios/sdk/WeexSDK/Sources/Component/WXWebComponent.m]: done!

**Pull request closed without merging**